### PR TITLE
docgen: implement doc link resolution in current module

### PR DESCRIPTION
**Pull request by a-mr**
First part of https://github.com/nim-lang/RFCs/issues/125#, smart doc links. Only local links (inside a given current module) are supported here.

It works by adding additional anchors from `docgen.nim`.

An example (modified from https://github.com/nim-lang/RFCs/issues/125#):

```nim
type SortOrder* = enum k0, k1
proc sortOrder*() = discard
proc foo*(a: int, sortOrder: SortOrder) = discard
proc foo*[T](a: int, sortOrder: SortOrder, c: T) = discard
proc foo*() =
  ## See: SortOrder_ (will link to type)
  ## See: sortOrder_ (will link to proc)
  ## See: foo_ (will link to the whole group of procs 'foo')
  ## See: `foo(int, SortOrder, T)`_ (will link to 2nd overload)
  ## See: `foo()`_ (will link to itself)                                                                                
```

image

Implementation throws warnings in case of ambiguities. In this case one needs to specify the reference more, e.g. `` `iterator split`_`` or  `` `proc split`_`` instead of just ``split_``.

A syntax note:
* one-word reference can be just suffixed with `_` without using enclosing backticks `` ` ``
* multiple-words references should be with backticks: `` `...`_``

TODO:
- [x] make it more WYSIWYG: 1) references like ``foo_`` and `` `proc foo`_`` should display as [foo proc]() 2) `` `foo(par1, par2)`_`` as [foo (par1, par2) proc]() 3) `` `foo(Type1, Type2)`_`` as [foo (Type1, Type2) proc]()
- [x] when there is only 1 overload:  show full name with function parameters (even if it's short-style link like foo_) UPDATE: only from tooltip on hover
- [x] when there is more than 1 overload for short-style links: show some descriptive link like proc foo (2 overloads)
- [x] add more tests outlined by @timotheecour
- [x] add test with inclusion of a module
- [x] check it in Latex backend
- [x] check some edge cases like the order dependent example from RFC: **currently not working:**
overloads with type constraints on generic parameters

cc @timotheecour

### CODE REVIEW

**Comment by timotheecour** [comment]
Excellent!
seems to work well and is intuitive. After playing with it a bit I found this bug (reduced from a weird behavior i was seeing after modifying compileOption links):

## bug 1
```nim
when defined case1: # D20210808T110300
  proc foo*(option: string): bool =
    ## * `foo <#foo,string,string>`_ (broken link, not expected to work)

  proc foo2*(option, arg: string): bool =
    ## * foo_ (BUG: doesn't show link to `foo`)

  proc foo3*(option: string): bool =
    ## * `foo3 <#foo,string,string>`_ (BUG2: doesn't link)

  proc foo3*(option, arg: string): bool =
    ## * foo3_ (BUG3: doesn't link)

```

## bug 2
```nim
when defined case2: # D20210808T110332
  proc compileOption5*(option: string): bool {.
    magic: "CompileOption", noSideEffect.} =
    ## See also:
    ## * `compileOption5 <#compileOption5,string,string>`_ for enum options

  proc compileOption5*(option, arg: string): bool {.
    magic: "CompileOptionArg", noSideEffect.} =
    ## See also:
    ## * compileOption5_ (BUG: should link to group)
```

## bug 3
```nim
when defined case3: # D20210808T110352
  proc foo*(option: string): bool =
    ## See also:
    ## * `foo_ <#foo,string,string>`_  (BUG: shows as ``foo_``)

  proc foo*(option, arg: string): bool =
    ## See also:
    ## * foo_ (BUG: should link to group)
```

they're probably all related

File: compiler/docgen.nim [hunk 1]
```diff
@@ -819,6 +821,18 @@ proc complexName(k: TSymKind, n: PNode, baseName: string): string =
       result.add(defaultParamSeparator)
       result.add(params)
 
+proc rstLinkName(k: TSymKind, n: PNode, baseName: string): string =
```
Comment (timotheecour) [inline compiler/docgen.nim]: links in lib/system_overview.rst (included from system.nim) work as intended, which is great; can you add a test to ensure links in includes keep working?

File: compiler/docgen.nim [hunk 2]
```diff
@@ -819,6 +821,18 @@ proc complexName(k: TSymKind, n: PNode, baseName: string): string =
...
+  ## Creates a link anchor in the format compatible with the corresponding
+  ## human-readable RST references.
```
Comment (timotheecour) [inline compiler/docgen.nim]: can you add tests with operators? eg:
```nim
## see `[]`_
## see `[]=`_
```

File: compiler/docgen.nim [hunk 3]
```diff
@@ -819,6 +821,18 @@ proc complexName(k: TSymKind, n: PNode, baseName: string): string =
...
+  result = baseName
```
Comment (timotheecour) [inline compiler/docgen.nim]: can you add tests with comma separated links, eg:
```nim
## see fn1_, fn2_, `fn3`_, `fn4`_
```

File: lib/packages/docutils/rst.nim [hunk 4]
```diff
@@ -2904,6 +2970,63 @@ proc preparePass2*(s: PRstSharedState, mainNode: PRstNode) =
...
+      if (i > 0 and s[i-1] in nimDeclPunctuation) or
+          (i+1 < s.len and s[i+1] in nimDeclPunctuation):
+        inc i
+      else:
+        result.add s[i]
+        inc i
+      newWord = true
+    of nimDeclPunctuation:
+      if afterSymbolName and s[i] == '[':
+        while i < s.len:  # skip type parameters in brackets
```
Comment (Araq) [inline lib/packages/docutils/rst.nim]: er... what? `proc f(x: G[int])` is not the same as `proc f(x: G[string])`

Comment (a-mr) [reply lib/packages/docutils/rst.nim]: it's just a bad wording, I changed it. I mean only "definition" brackets right after symbol name and before opening `(`.

I also added test case to `utils.nim` to ensure it works fine.

**Comment by a-mr** [comment]
@timotheecour 

In all those bugs you hit a (mis)feature of RST. A link like `` `foo <site#anchor>`_`` generates a link alias `foo`. When you use `foo_` it selects this manually set variant because it has priority. So this is as expected. In the new version a warning is emitted about this conflict with line info where all these variants were originally defined. May be I need to go even further and add to the warnings a phrase that these link aliases were implicitly generated.

Also note that handling of these links `` `foo <...>`_`` were not changed in this PR.
Generally speaking, they are  manual and not checked and when you are using them "you are on your own". After this PR and its follow-ups we are not supposed to even use these manual links for linking to symbols, though they will be still relevant for referencing external sites and whole Nim modules (though we can add syntax & automatic resolution even to the latter case at some stage too).

**Comment by a-mr** [comment]
Marked as WIP as it's a big change and there is no need to rush.

Current list of TODOs:

- [x] when there is only 1 overload: ~~Don't generate groups~~ (it's better to generate groups anyway to avoid breaking external links if e.g. number of overloads goes 2 -> 1)  and show full name with function parameters (even if it's short-style link like ``foo_``)
- [x] when there is more than 1 overload for short-style links: show some descriptive link like [proc foo (2 overloads)]()
- [x] add more tests outlined by @timotheecour 
- [x] check it in Latex backend
- [x] check some edge cases like the order dependent example from RFC:
      **currently not working**
```nim
      ## ref fn_ (OK) and `fn(T)`_ (COLLISION) and `fn(T_2)`_ (NOT FOUND)
      proc fn*[T: SomeFloat](a: T)=discard
      proc fn*[T: SomeInteger](a: T)=discard

```

**Comment by timotheecour** [comment]
> n all those bugs you hit a (mis)feature of RST. A link like `foo <site#anchor>`_ generates a link alias foo. When you use foo_ it selects this manually set variant because it has priority. So this is as expected

I see, i thought it would only specify the display text as `foo`, didn't realize it'd also create a link alias `foo`; is that something we want to preserve? but fine, if we stick to the "new form" this isn't a problem (but it is a problem if mixing new form and old form as shown in https://github.com/nim-lang/Nim/pull/18642#issuecomment-893660834

What does it mean for existing moduels that use the existing syntax `foo_ <#foo,string,string>`_  ? can we upgrade them 1 by one to use instead `foo_` or we also need to make sure other modules referencing it use `foo_`?


> In the new version a warning is emitted about this conflict with line info where all these variants were originally defined.

it helps, thanks

> May be I need to go even further and add to the warnings a phrase that these link aliases were implicitly generated.

maybe

**Comment by a-mr** [comment]
> What does it mean for existing moduels that use the existing syntax foo_ <#foo,string,string>_ ? can we upgrade them 1 by one to use instead foo_ or we also need to make sure other modules referencing it use foo_?

Yes, we can and actually updating a module as whole  eliminates a possibility of collision. Also warning will necessarily be generated so it will be impossibly to forget anything. And besides current links are almost always input as `` `parseCmdLine proc  <#parseCmdLine,string>`_``, w.r.t. to our new syntax it has a reverse of `parseCmdLine` and `proc` hence there is no collision.

And problems when referencing between modules are impossible because the format of anchors has not changed. All in all, there is almost no risk in this regard.

**Comment by Varriount** [comment]
Just want to say this is very impressive! Thanks for your work!

File: nimdoc/testproject/subdir/subdir_b/utils.nim [hunk 5]
```diff
@@ -115,3 +115,8 @@ proc f*(x: G[int]) =
 proc f*(x: G[string]) =
   ## See also `f(G[int])`_.
   discard
+
+## Ref. `[]`_ is the same as `proc \`[]\`(G[T])`_ because there are no
```
Comment (timotheecour) [inline nimdoc/testproject/subdir/subdir_b/utils.nim:129]: would this work to avoid quoting?
```nim
## See `proc \`$\`(int)`_
## See ``proc `$`(int)``_
```
the 2nd is more readable because it's easier to copy paste from a declaration

File: nimdoc/testproject/subdir/subdir_b/utils.nim [hunk 6]
```diff
@@ -115,3 +115,8 @@ proc f*(x: G[int]) =
...
+## overloads.
+
```
Comment (timotheecour) [inline nimdoc/testproject/subdir/subdir_b/utils.nim:131]: can you add more tests especially for edge cases, eg:
```
## See `$`_, `[]=`_
## See `'big`_ (for user defined literals)
## See `foo_bar_` (for `foo_bar`)
```

File: nimdoc/testproject/subdir/subdir_b/utils.nim (hunk 5 above)
Comment (a-mr) [reply nimdoc/testproject/subdir/subdir_b/utils.nim:129]: Not now :-(

In original RST it would not even require escaping, but our implementation was tweaked in favor of Markdown-like inline markup rules, which allow to put the ending backtick anywhere and also put starting backtick anywhere. So with single-backtick syntax we are bound to escaping of backticks.

Regarding double-backticks: the syntax is sane and I like it but it's not standard and we do not support it, though it would probably be easy to implement. I think anything like this will be left for future PRs anyway to avoid conflation here.

I agree it's ugly, that's why to make it less painful I added possibility to ref. operators the short way `` `$`_`` (instead of a long way like
```
`\`$\``_
```
)
but for any compound anchors one still has to use escaping:
```
`proc \`$\``_
```

**Comment by Varriount** [comment]
Is this still a WIP, or is it ready for a full review?

**Comment by a-mr** [comment]
> Is this still a WIP, or is it ready for a full review?

WIP.

The implementation can still change significantly to add more flexibility ("WYSIWIG") and fix a bug with type constraints.
Also the current scheme of adding aliases may cause memory consumption, I need to do a simple benchmark at least.

**Comment by Varriount** [comment]
Ok, I've converted this PR to a draft - when it's ready for a full review, you can convert it back and modify the title to remove the "WIP".

File: doc/docgen.rst [hunk 7]
```diff
@@ -229,6 +229,146 @@ Output::
 Note that the `jsondoc`:option: command outputs its JSON without pretty-printing it,
 while `jsondoc0`:option: outputs pretty-printed JSON.
 
+
+Referencing Nim symbols: simple documentation links
+===================================================
+
+You can reference Nim identifiers from Nim documentation comments (currently
+only inside one ``.nim`` file).
```
Comment (timotheecour) [inline doc/docgen.rst]: maybe also mention that it works for nim files including an rst file?

File: doc/docgen.rst [hunk 8]
```diff
@@ -229,6 +229,146 @@ Output::
...
+
+   proc foo*(a: int, b: float): string
+
+Generally following syntax is allowed for referencing `foo`:
+
+*  short (without parameters):
+
+   A. non-qualified::
+
+        foo_
```
Comment (timotheecour) [inline doc/docgen.rst:275]: maybe mention this is the preferred form because:
* otherwise ppl will get confused as to which form to use by default
* simpler = better
* more robust to API changes (eg proc => template, adding optional params etc)

(EDIT: i see you mention it later, but might as well mention it here too)

File: doc/docgen.rst [hunk 9]
```diff
@@ -229,6 +229,146 @@ Output::
...
+
+   A short form works without additional backticks::
+
+     `[]`_
+     `$`_
+
+   However for fully-qualified reference copy-pasting backticks (`) into other
+   backticks will not work in our RST parser (because we use Markdown-like
+   inline markup rules). In the case of `$` it's allowed to delete backticks,
+   but in case of `[]` one needs to keep backticks and escape them with
```
Comment (timotheecour) [inline doc/docgen.rst]: that's a bit nasty, can we avoid this?
this should work and not be ambiguous:
```nim
`func []`_ or, in detail:
`func [][T](x: openArray[T]): T`_
```

File: doc/docgen.rst [hunk 10]
```diff
@@ -229,6 +229,146 @@ Output::
...
+   with higher priority is selected. Manual anchors have highest
+   priority, then go automatic RST anchors; then Nim-generated anchors
+   (while procs have higher priority than other Nim symbol kinds).
+
+.. Note:: A bit special case is operators
+   (as their signature is also defined with `\``):
+
+   .. code:: Nim
+
+      func `[]`*[T](x: openArray[T]): T
```
Comment (timotheecour) [inline doc/docgen.rst]: should the export marker `*` be present?

Comment (a-mr) [reply doc/docgen.rst]: yes. I mentioned it above:

>This pertains to any exported symbol

Non-exported symbols don't get into documentation (by default, without `--docInternal`) and hence should not be referenced from module description or exported symbols doc comments.

I'm planning to propose a special section `.. internal::` (RST directive) inside doc comments that would be allowed to contain references to non-exported symbols. I expect to send a separate RFC for discussion about literate programming in Nim after finishing this PR and this RFC will contain that feature as its 1st part.

File: doc/docgen.rst (hunk 9 above)
Comment (a-mr) [reply doc/docgen.rst]: yes, at first glance I don't see any principal obstacle. It may well be a bug in my working copy :-)

File: doc/docgen.rst (hunk 10 above)
Comment (timotheecour) [reply doc/docgen.rst]: sorry ignore my comment, i though you had added the `*` as part of the link

File: doc/docgen.rst [hunk 11]
```diff
@@ -282,6 +285,8 @@ Generally following syntax is allowed for referencing `foo`:
 
           `foo(a, b)`_
 
+         .. _`form A2`:
```
Comment (timotheecour) [inline doc/docgen.rst]: I don't understand this

File: doc/docgen.rst [hunk 12]
```diff
@@ -299,36 +304,14 @@ Generally following syntax is allowed for referencing `foo`:
...
+   one of shorter forms::
 
      binarySearch_
-     `proc binarySearch(a, key, cmp)`_
+     `binarySearch(a, key, cmp)`_
 
-   If you use a short form and have an ambiguity then just add some
-   additional info.
+   Brevity is better for reading! If you use a short form and have an
+   ambiguity problem (see below) then just add some additional info.
```
Comment (timotheecour) [inline doc/docgen.rst:313]: how about adding something like this:

Furthermore, the simplest form (`binarySearch_`) is usually preferable because it will generate links that remain valid even if underlying API is modified (e.g. via added optional params, or changing the routine kind).

File: doc/docgen.rst [hunk 13]
```diff
@@ -346,28 +329,56 @@ recognized fine::
...
+                               cmp: proc (x: T; y: K): int {.closure.}): int
+                                          ~~    ~~   ~~~~~
+
+   you cannot use names underlined by `~~` so it must be referenced with
+   ``cmp: proc(T, K)``. Hence these forms are valid::
+
+     `binarySearch(a: openArray[T], key: K, cmp: proc(T, K))`_
+     `binarySearch(openArray[T], K, proc(T, K))`_
+     `binarySearch(a, key, cmp)`_
+2. default values in routine parameters are not recognized, one need to
```
Comment (timotheecour) [inline doc/docgen.rst]: need => needs

File: doc/docgen.rst [hunk 14]
```diff
@@ -346,28 +329,56 @@ recognized fine::
...
+   specify the type instead. E.g. for referencing `proc f(x = 7)` use::
```
Comment (timotheecour) [inline doc/docgen.rst]: can you add a TODO for future work somewhere? fixing this would be a nice improvement

File: doc/docgen.rst [hunk 15]
```diff
@@ -346,28 +329,56 @@ recognized fine::
...
+   inline markup rules). You need either to delete backticks or keep
+   them and escape with backslash \\::
 
-      `func \`[]\``_  or, in detail:
-      `func \`[]\`*[T](x: openArray[T]): T`_
-      `func $`_
+      no backticks: `func $`_
+      escaped:  `func \`$\``_
+      no backticks: `func [][T](x: openArray[T]): T`_
+      escaped:  `func \`[]\`[T](x: openArray[T]): T`_
```
Comment (timotheecour) [inline doc/docgen.rst:400]: do we need both? honest question; i feel like just using the no backticks form without giving an option would be best, and simplest to use.

Comment (a-mr) [reply doc/docgen.rst:400]: yes, we do need both. One can want for the link to look syntactically correct (e.g. for educational reasons).

Comment (timotheecour) [reply doc/docgen.rst:400]: the ideal link would be a copy-paste of the declaration, possible non-ambiguous ways could be:
```nim
## see ` func `$`(a: int): string `_
## see ``func `$`(a: int): string``_
## see ```func `$`(a: int): string```_
```

(at expense of rst spec, which we already violate)

then we'd need only 1 way, and avoid having to give a choice with 2 sub-optimal ways

File: compiler/docgen.nim (hunk 1 above)
Comment (a-mr) [reply compiler/docgen.nim]: ok, done, thank you for the idea!

File: compiler/docgen.nim (hunk 2 above)
Comment (a-mr) [reply compiler/docgen.nim]: done, tests added.

File: compiler/docgen.nim (hunk 3 above)
Comment (a-mr) [reply compiler/docgen.nim]: done.

File: nimdoc/testproject/subdir/subdir_b/utils.nim (hunk 6 above)
Comment (a-mr) [reply nimdoc/testproject/subdir/subdir_b/utils.nim:131]: added. Thanks for testcases, they helped to catch some bugs.

File: doc/docgen.rst (hunk 7 above)
Comment (a-mr) [reply doc/docgen.rst]: ok, mentioned.

File: doc/docgen.rst (hunk 8 above)
Comment (a-mr) [reply doc/docgen.rst:275]: i think one suggestion is enough. Besides some people may like to provide parameters to procs — I would prefer that in most cases also.

File: doc/docgen.rst (hunk 9 above)
Comment (a-mr) [reply doc/docgen.rst]: it's working now.

File: doc/docgen.rst (hunk 11 above)
Comment (a-mr) [reply doc/docgen.rst]: that was an unsuccessful to set an anchor at bullet point, there is some bug here unrelated with this PR.

File: doc/docgen.rst (hunk 13 above)
Comment (a-mr) [reply doc/docgen.rst]: fixed.

File: doc/docgen.rst (hunk 14 above)
Comment (a-mr) [reply doc/docgen.rst]: the current implementation does not use Nim compiler facilities to parse all these links, so to fix that for all possible cases would be impossible. The only solution would be to re-write this implementation which is not in my plans.

File: doc/docgen.rst (hunk 15 above)
Comment (a-mr) [reply doc/docgen.rst:400]: I see, it's roughly the same as https://github.com/nim-lang/RFCs/issues/355.

this syntax is also not perfect, it does not allow to start from a backtick. While our current syntax does:

```
`\`$\`(a: int)`_
```
(Correction: one can input
```
`` `$`(a: int) ``_
```
though, it's just that further tweaking of RST inline markup rules in favor of Markdown is needed hear.)

But I agree that such syntax would be good... It's just not among first priorities...

**Comment by a-mr** [comment]
@Araq @timotheecour @narimiran 

What are you plans for review & conclusion?
Do you have any time during a month? I want to continue working on docgen but this PR kind of hinders me because quite a bit of refactoring was done here, it would be unpleasant to merge that afterwards. If you want we can have a remote session e.g. by Skype discussing this PR.

The current implementation fixes all the problems that were discovered.
It's different from the original code. Original one made a string normalization for both generated anchor and reference and did a simple hash table search. The new one parses a reference ( in `dochelpers.nim`) so that all parameters (including generics) can be compared with an anchor. Originally I thought about re-using parser code from the compiler but abandoned this idea because:
* only signatures need to be supported in `rst.nim`
* also signatures that don't abide by Nim rules are supported, e.g. one can write `walkDir(d) iterator` — without type and with `iterator` at the end. This originated from the existing practice of naming links in Nim's repo.

**Comment by Araq** [comment]
Sorry for the delay, we're reviewing it this week.

**Comment by narimiran** [comment]
It is hard to do a detailed review of a large PR, but I have tested the changes proposed in this PR by building the docs for `strutils` (while changing the links to the new short form) and I like the new behavior, e.g. grouping of `split` funcs when you link to (one of) them.

I'm merging this now; we have plenty of time until next stable version (1.8.0) to iron out any quirks/bugs that might come up which we haven't considered now.

**Comment by a-mr** [comment]
OK, thx,  so I will try to convert all links e.g. in `os.nim`.

Also the next part of work will be cross-module link support. I hope it will be ready by 1.8.0 also.
